### PR TITLE
Added method overloading for Xamarin.Android

### LINKED ISSUE #24: Improve getting the current context
## Summary
Getting the current activity is not done quite correctly. 

```csharp
public static partial class Platform
{
    static Activity currentActivity;

    internal static Activity AppActivity
         => currentActivity
        ?? throw ExeptionHelper.ActivityNotDetected;
}
```
[Permalink](https://github.com/dimonovdd/Xamarin.MediaGallery/blob/85d02822b3d513b3784b0d6dd8d14d59f6b71cf7/MediaGallery/Platform/Platform.android.cs#L11)

Maybe we could replace this with a `IActivityLifecycleCallbacks` or the `Android.App.Application.Context`

### PR DESCRIPTION
**Pull request by dimonovdd**
### Description
Added methods overloading for Xamarin.Android with `activity` parameter

### Related to issue
- Closes #24 

### API Changes
**Added:**
```csharp
public static Task<IEnumerable<IMediaFile>> PickAsync(Activity activity, MediaPickRequest request, CancellationToken token = default);

public static async Task SaveAsync(Activity activity, MediaFileType type, byte[] data, string fileName);

public static async Task SaveAsync(Activity activity, MediaFileType type, string filePath);

public static async Task SaveAsync(Activity activity, MediaFileType type, Stream fileStream, string fileName);
```

### PR Checklist ###

- [ ] All projects build
- [ ] Has samples
- [ ] Rebased onto current `main`

### CODE REVIEW

**Comment by dimonovdd** [comment]
@pictos @ahoefling  I'm thinking of another way:

```csharp
public static partial class Platform
{
    static Func<Activity> getCurrentActivity;

    public static void Init(Func<Activity> getCurrentActivity, Bundle bundle)
         => getCurrentActivity = getCurrentActivity;
        
    internal static Activity AppActivity
        => getCurrentActivity?.Invoke()
            ?? throw ExceptionHelper.ActivityNotDetected;
}
```

File: MediaGallery/MediaGallery/SaveMedia.android.cs [hunk 1]
```diff
@@ -20,23 +21,65 @@ namespace NativeMedia
...
+        }
+
+        public static async Task SaveAsync(Activity activity, MediaFileType type, Stream fileStream, string fileName)
+        {
+            await CheckPossibilitySave();
+            if (fileStream == null)
+                throw new ArgumentNullException(nameof(fileStream));
+            CheckFileName(fileName);
+            if(activity == null)
+                throw new ArgumentNullException(nameof(activity));
```
Comment (pictos) [inline MediaGallery/MediaGallery/SaveMedia.android.cs:54]: I saw this null check in a lot of places, you can create a helper method to check it for you

**Review by pictos — COMMENTED** [review]
I can't reproduce the issue in this branch... Maybe I can do it in main branch? If don't how about create a TestCase folder in sample app and add the failing scenario

**Comment by pictos** [comment]
> @pictos @ahoefling I'm thinking of another way:
> 
> ```cs
> public static partial class Platform
> {
>     static Func<Activity> getCurrentActivity;
> 
>     public static void Init(Func<Activity> getCurrentActivity, Bundle bundle)
>          => getCurrentActivity = getCurrentActivity;
>         
>     internal static Activity AppActivity
>         => getCurrentActivity?.Invoke()
>             ?? throw ExceptionHelper.ActivityNotDetected;
> }
> ```

Not sure if you need to go with a fancy Func for that... In Forms apps, the Activity will be always the same

**Comment by pictos** [comment]
Also, you have a XE dependency, so you don't need to capture the Activity in your Platform class

**Comment by dimonovdd** [comment]
@pictos This issue is mainly related to xamarin.Android apps

**Comment by pictos** [comment]
@dimonovdd 
> @pictos This issue is mainly related to xamarin.Android apps

I see, in that case, I believe that you can use the XE to grab the current activity since that implements all lifecycle

**Comment by dimonovdd** [comment]
I don't quite understand what you mean

**Comment by pictos** [comment]
@dimonovdd your lib has a reference to Xamarin.Essentials. So instead of collecting the activity let the XE handle that for you... I did this on your code and worked.

```csharp
// on your lib
public static void Init(Activity activity, Bundle bundle) { }
 internal static Activity AppActivity => Xamarin.Essentials.Platform.CurrentActivity;

// Sample
base.OnCreate(savedInstanceState);
//NativeMedia.Platform.Init(this, savedInstanceState);
Xamarin.Essentials.Platform.Init(this, savedInstanceState);
global::Xamarin.Forms.Forms.Init(this, savedInstanceState);
```

**Comment by dimonovdd** [comment]
@pictos  Thank you very much. I completely forgot about the existence of `Xamarin.Essentials.Platform.CurrentActivity` property.

**Comment by dimonovdd** [comment]
@pictos Thank you very much. I completely forgot about the existence of Xamarin.Essentials.Platform.CurrentActivity property.

**Pull request closed without merging**